### PR TITLE
A task that will connect to the SQL database, fetch data from all tables and create a Cortex Dataset

### PR DESCRIPTION
**Pull request by duskovesic**

### CODE REVIEW

**Review by VukManojlovic — CHANGES_REQUESTED** [review]

File: tasks/sql-connector/main.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+
+def fetchAllData(conn: Any, dataset: CustomDataset, queryGetTables: str, queryGetRows: str) -> None:
+    cursor = conn.cursor()
+    cursor.execute(queryGetTables)
+    tables = cursor.fetchall()
+    tables = [table[0] for table in tables]
+
+    for table in tables:
+        tableData = []
```
Comment (VukManojlovic) [inline tasks/sql-connector/main.py:45]: When defining empty lists and dictionaires typing should be added.
For example: `listVariable: list[str] = []`

**Review by VukManojlovic — CHANGES_REQUESTED** [review]

File: tasks/sql-connector/main.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        for row in rows:
+            tableData.append(dict(zip(columnNames, list(row))))
+
+        sampleNameCSV = f"{table}.csv"
+        with open(sampleNameCSV, "w", newline="") as file:
+            writer = csv.DictWriter(file, fieldnames = columnNames)
+            writer.writeheader()
+            writer.writerows(tableData)
+
+        sampleNameZIP = f"{table}.zip"
```
Comment (dule1322) [inline tasks/sql-connector/main.py:63]: We use camelCase style so `sampleNameCSV` -> `sampleNameCsv`, `sampleNameZIP` -> `sampleNameZip`

File: tasks/sql-connector/main.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    if databaseType == "PostgreSQL":
+        conn = connectPostgresqlDatabase(connectionConfig)
+
+        if conn:
+            dataset = CustomDataset.createDataset(f"{taskRun.id}-{database}", taskRun.projectId)
+            queryGetTables = f"SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'"
+            queryGetRows = f"SELECT column_name FROM information_schema.columns WHERE table_schema = 'public' AND table_name = "
+            fetchAllData(conn, dataset, queryGetTables, queryGetRows)
+
```
Comment (dule1322) [inline tasks/sql-connector/main.py:112]: Remove this newline

File: tasks/sql-connector/main.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        if conn.is_connected():
+            dataset = CustomDataset.createDataset(f"{taskRun.id}-{database}", taskRun.projectId)
+            queryGetTables = f"SELECT table_name FROM information_schema.tables WHERE table_schema = '{database}'"
+            queryGetRows = f"SELECT column_name FROM information_schema.columns WHERE table_schema = '{database}' AND table_name = "
+            fetchAllData(conn, dataset, queryGetTables, queryGetRows)
+
+        else:
+            logging.warning(">> [SQL Connector] Problem with the database connection")
+
+    if databaseType == "PostgreSQL":
```
Comment (dule1322) [inline tasks/sql-connector/main.py:104]: This should be `elif`

File: tasks/sql-connector/main.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    if databaseType == "MySQL":
+        conn: Union[CMySQLConnection, connection] = connectMysqlDatabase(connectionConfig)
+
+        if conn.is_connected():
+            dataset = CustomDataset.createDataset(f"{taskRun.id}-{database}", taskRun.projectId)
+            queryGetTables = f"SELECT table_name FROM information_schema.tables WHERE table_schema = '{database}'"
+            queryGetRows = f"SELECT column_name FROM information_schema.columns WHERE table_schema = '{database}' AND table_name = "
+            fetchAllData(conn, dataset, queryGetTables, queryGetRows)
+
```
Comment (dule1322) [inline tasks/sql-connector/main.py:100]: Remove this newline

File: tasks/sql-connector/task.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+id: sql-connector
+name: SQL Connector
```
Comment (dule1322) [inline tasks/sql-connector/task.yaml:2]: This should be called `transfer-data-from-sql-database`, or `upload-data-from-sql-database`. Also, `id` and `name` must match, we will be removing `name` completely soon, so keep them identical until we do.

File: tasks/sql-connector/task.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+description: A task that will connect to the SQL database using credentials, fetch data from all tables, and insert it into a Cortex Dataset
+is_active: true
+project_type:
+param_groups:
+  - name: inputs
+    params:
+      - name: database_type
```
Comment (dule1322) [inline tasks/sql-connector/task.yaml:9]: This name should be camelCase

File: tasks/sql-connector/main.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    try:
+        conn = psycopg2.connect(**connectionConfig)
+    except psycopg2._psycopg.Error as e:
+        logging.error(f">> [SQL Connector] Error while connecting to database: {e}")
+
+    return conn
+
+
+def fetchAllData(conn: Any, dataset: CustomDataset, queryGetTables: str, queryGetRows: str) -> None:
```
Comment (dule1322) [inline tasks/sql-connector/main.py:38]: Why is `conn` of type `Any`? `Any` should almost never be used.

File: tasks/sql-connector/main.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    cursor = conn.cursor()
+    cursor.execute(queryGetTables)
+    tables = cursor.fetchall()
+    tables = [table[0] for table in tables]
```
Comment (dule1322) [inline tasks/sql-connector/main.py:42]: Is `table[0]` guaranteed to not raise an index-out-of-bounds errror?

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: tasks/sql-connector/main.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+    return conn
+
+
+def fetchAllData(conn: Union[CMySQLConnection, connection], dataset: CustomDataset, queryGetTables: str, queryGetRows: str) -> None:
+    cursor = conn.cursor()
+    cursor.execute(queryGetTables)
+    tables = cursor.fetchall()
+    if len(tables) < 1:
+        raise RuntimeError("In the database, there are no tables")
```
Comment (dule1322) [inline tasks/sql-connector/main.py:43]: I would change this to something like `There are no tables in the database`.

File: tasks/sql-connector/main.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        columnNames = [name[0] for name in columnNames]
+
+        cursor.execute(f"SELECT * FROM {table}")
+        rows = list(cursor.fetchall())
+
+        for row in rows:
+            tableData.append(dict(zip(columnNames, list(row))))
+
+        sampleNameCsv = f"{table}.csv"
+        with open(sampleNameCsv, "w", newline="") as file:
```
Comment (dule1322) [inline tasks/sql-connector/main.py:61]: There should be whitespace around `=` operator

File: tasks/sql-connector/main.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        conn: Union[CMySQLConnection, connection] = connectMysqlDatabase(connectionConfig)
+
+        if conn.is_connected():
+            dataset = CustomDataset.createDataset(f"{taskRun.id}-{database}", taskRun.projectId)
+            queryGetTables = f"SELECT table_name FROM information_schema.tables WHERE table_schema = '{database}'"
+            queryGetRows = f"SELECT column_name FROM information_schema.columns WHERE table_schema = '{database}' AND table_name = "
+            fetchAllData(conn, dataset, queryGetTables, queryGetRows)
+        else:
+            logging.warning(">> [SQL Connector] Problem with the database connection")
+
```
Comment (dule1322) [inline tasks/sql-connector/main.py:104]: Remove newline

**Review by dule1322 — CHANGES_REQUESTED** [review]

**Review by dule1322 — APPROVED** [review]

**Pull request merged**